certmetrics/metrics.go

```diff
@@ -0,0 +1,51 @@
+// Package certmetrics will be used to register and emit metrics for certificates in memory
+package certmetrics
+
+import (
+	"crypto/x509"
+	"sort"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+var certificateExpirationTimes = promauto.NewGaugeVec(
+	prometheus.GaugeOpts{
+		Name: "certificate_expiration_timestamp_seconds",
+		Help: "Expiration times of gokeyless certs",
+	},
+	[]string{"serial_no", "cn", "hostnames", "ca", "server", "client"},
+)
+
+// Observe takes in a list of certs and emits its expiration times
+func Observe(certs ...*x509.Certificate) {
+	for _, cert := range certs {
+		hostnames := cert.DNSNames
+		sort.Strings(hostnames)
```

Comment:
This is sorting `cert.DNSNames` in place, it's not a copy: https://goplay.space/#EOmjxbq7se2.

This could cause data races depending on how this function is used. I think it's best to skip the sort here, as the order is already deterministic based on the cert contents.

Reply:
idk how i missed it, i expected this to be references and then tested to ensure. But it turned out to be a copy so kept this, clearly i was sleeping.